src/lib/util/LayerManager.ts

```diff
@@ -0,0 +1,162 @@
+    this.unregister = this.unregister.bind(this);
+    this.redraw = this.redraw.bind(this);
+    this.resize = this.resize.bind(this);
+    this.getRenderingLayerId = this.getRenderingLayerId.bind(this);
+
+    this.currentLayerId = 1;
+    this.setups = new Map();
+    this.renderers = new Map();
+    this.dispatchers = new Map();
+
+    this.needsSetup = false;
+    this.needsResize = true;
+    this.needsRedraw = true;
+
+    this.renderingLayerId = 0;
+    this.activeLayerId = 0;
+
+    this.layerSequence = [];
+  }
+
+  redraw() {
+    this.needsRedraw = true;
+  }
+
+  resize() {
+    this.needsResize = true;
+    this.needsRedraw = true;
+  }
+
+  register({
+    setup,
+    render,
+    dispatcher
+  }: {
+    setup: Render | undefined;
+    render: Render;
+    dispatcher: LayerEventDispatcher;
+  }) {
+    if (setup) {
+      this.setups.set(this.currentLayerId, setup);
+      this.needsSetup = true;
+    }
+
+    this.renderers.set(this.currentLayerId, render);
+
+    this.dispatchers.set(this.currentLayerId, dispatcher);
+
+    this.needsRedraw = true;
+    return this.currentLayerId++;
+  }
+
+  unregister(layerId: number) {
+    this.renderers.delete(layerId);
+    this.dispatchers.delete(layerId);
+    this.needsRedraw = true;
+  }
+
+  render({
+    autoclear,
+    pixelRatio,
+    context,
+    width,
+    height
+  }: {
+    width: number;
+    height: number;
+    context: CanvasRenderingContext2D;
+    pixelRatio: number;
+    autoclear: boolean;
+  }) {
+    const renderProps = { context, width, height };
+
+    if (this.needsResize) {
+      context.setTransform(pixelRatio, 0, 0, pixelRatio, 0, 0);
+      this.needsResize = false;
+    }
+
+    if (this.needsSetup) {
+      for (const [layerId, setup] of this.setups) {
+        setup(renderProps);
+        this.setups.delete(layerId);
+      }
+
+      this.needsSetup = false;
+    }
+
+    if (this.needsRedraw) {
+      if (autoclear) {
+        context.clearRect(0, 0, width, height);
+      }
+
+      for (const layerId of this.layerSequence) {
+        this.renderingLayerId = layerId;
+        this.renderers.get(layerId)?.(renderProps);
+      }
+
+      this.needsRedraw = false;
+    }
+  }
+
+  setActiveLayer(layer: number, e: MouseEvent | TouchEvent) {
+    if (this.activeLayerId === layer) return;
+
+    if (e instanceof MouseEvent) {
+      this.dispatchLayerEvent(new PointerEvent('pointerleave', e));
+      this.dispatchLayerEvent(new MouseEvent('mouseleave', e));
+    }
+
+    this.activeLayerId = layer;
+    this.activeLayerDispatcher = this.dispatchers.get(layer);
+
+    if (e instanceof MouseEvent) {
+      this.dispatchLayerEvent(new PointerEvent('pointerenter', e));
+      this.dispatchLayerEvent(new MouseEvent('mouseenter', e));
+    }
+  }
+
+  dispatchLayerEvent(e: MouseEvent | TouchEvent) {
+    if (!this.activeLayerDispatcher) return;
+
+    let detail: LayerEventDetail = {};
+
+    if (window.TouchEvent && e instanceof TouchEvent && e.touches.length) {
```

Comment:
For the null check, LayerEventDetail will never have null x and y values.

From [w3c Touch Events](https://www.w3.org/TR/touch-events/#attributes-2):
> The Touch Events specification defines a set of low-level events that represent **one or more points** of contact with a touch-sensitive surface

So here you can just set it to `LayerEventDetail | undefined` and set the JSON structure in each if statement accordingly. In fact, it may be better to remove the `e.touches.length` for less confusing error messages in the rare case that (if ever) `e.touches.length` is 0.

Reply:
I put this check (and the optional detail x and y values) in place specifically for `touchend` events, where `e.touches` is empty. But I looked into it more and `e.changedTouches` is populated for all three touch events, so I'll use that instead and include x and y values in every `detail`.